### ISSUE TITLE
Layer models

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#1353]
## Discussion

There have been numerous discussions lately of our layer models, and various refactors at play in #1325, #1040, #1343 and more. There seems to be universal agreement that having the dims model which contains slicing information and having the sliced data on each layer is a problem and will prevent things like orthoviews and multicanvas features more broadly. With that in mind I'd like to make the following proposal for after the refactor is complete.

**After refactors layer model should have no knowledge of:**
- Where the sliders are
- Where the camera is
- Where the mouse cursor is
- What part of the data has been sliced
- What part of the data is being viewed
- What part of the data has been clicked on or hovered over
- The thumbnail

**It might also be nice if the layer model also had no knowledge of:**
- How the user currently wants to interact with the data (i.e. which “mode”)
- Any data that is currently selected or highlighted (i.e. in shapes/ points)
- If the layer is “selected” / “visible” / “editable”

Note “visible” is currently exposed in main add_layer API and can be used to make loading lazy, so we might want to keep that or at least deprecate it slowly.

**The layer model should know:**
- The full nD data
- How the data is placed in a physical/ world coordinate system
- Properties of the data (like class labels for points or shapes)
- How the data would look after being rendered (colormaps, colors, opacity etc)

The `add_layer` API and `layer_state` serialization method should allow users access to all these properties and be everything a users needs to know what data is, where in space it is, and what it would look like if rendered.

**Methods of the layer model should allow:**
- Slicing
- Interactivity
- Viewing

And should take slice information, cursor information, and camera information as needed to produce the desired outputs. The exact interactivity requested will depending on what "mode" the rest of the app is in.

Feedback welcome!

**Comment by sofroniewn** [comment]
Here is a schematic of some of our current architecture in terms of properties and inheritance. Lots of these things are also coupled through out event system and not everything is here, but having this might aid some conversation 
![image](https://user-images.githubusercontent.com/6531703/84580038-2ff58b00-ad88-11ea-867c-e79ed0f58395.png)

**Comment by sofroniewn** [comment]
Focusing just on the `ViewerModel` side of things and thinking only about all the `Model` code we need (including for a world where we support multiple canvases), one reorganization could be as follows:

![image](https://user-images.githubusercontent.com/6531703/84583637-64307200-adaf-11ea-856f-29583f85d97a.png)

Where if there were 3 `Canvases` and 7 `Layers` you'd have 21 `SlicedData`. You'd always have 1 `Dims` for the sliders at the bottom of the Viewer. The 3 `SliceModels` would additionally contain information about which dimensions are being viewed in each slice. The `Controls` would now actually just be one model per layer type and represent state like what mode the controls (box in top left) were in and would only change when the type of layer active changed. The `Highlighting` would maybe just be one thing that contained state with what was being selected within a layer and what needed to be highlighted.

One would then need a controller to make sure all of these pieces fit together properly, and then our various views would need various pieces of these models to display things properly.

I'm sure there are many many flaws with the above diagram, but things I like are that a plugin developer that is purely interested in "Analysis" plugins which take in data operate on it can really just think about our "LayerModels" and `LayersList`. If they also want to think about what data is currently being viewed or where the cursor is they can use the `Dims` and `CanvasModels`. I'd need to think more about what it would mean for a plugin to access the information in `Highlighting` or `Controls`.

The last major thing this architecture is missing is any representation of supplemental visualizations/ 1-D plots, scatter plots, dendrograms, histograms etc. Probably ok to continue to defer discussion about them, but I don't want us to forget about them

**Comment by jni** [comment]
Thanks @sofroniewn! To me an interesting part of this proposal is that it seems like the amount of information being held by the layers could (almost?) be captured in a bare xarray! There is *one* more critical thing that the layer models hold, btw — their type!

Anyway, I totally agree that this is a necessary change in our data models, but we should think about this very carefully, because we should try to get this done in one go. A change of this magnitude will affect all dependent packages!

But, I really love the direction here.

**Comment by sofroniewn** [comment]
> To me an interesting part of this proposal is that it seems like the amount of information being held by the layers could (almost?) be captured in a bare xarray! 

Yes, whether we use an xarray or not is up for debate, but I want it to be very close, and for the serialization to effectively give us what we need for an xarray.

@kne42 and @0x00b1 might also want to look at the above proposal as they thing about how the controls/ event handler reactors in #1040 and #1325 are going

**Comment by kne42** [comment]
I really like the ideas you have here @sofroniewn! I think with regard to xarray, not all data will necessarily be array-based (I think?) so it would be kind of weird for xarray to be used as an integral part of the implementation.

**Comment by jni** [comment]
This seems like as good a place as any to note that ImageSlice and ImageView are, imho, not descriptive enough of what is going on. See @pwinston's description here for details:

https://github.com/napari/napari/pull/1565#discussion_r474566605

**Comment by jni** [comment]
Another note on this. Today @tlambert03 and I played around with resurrecting #838. It's not trivial! The main issue we were running into is that, until now, `Image.raw_to_displayed()` and `Image.rgb` never changed after instantiation. So, now, as opposed to six months ago, when we try to refresh and grab the slice *with new parameters*, the chunk loader says, you already have that slice, here you go, and returns the existing slice with the existing transformation and rgb state. (This is fine and expected behaviour given the architecture of the Image class until now.)

So, we appear to need some room somewhere between ImageSlice and ImageView, or maybe this is ImageView itself, for a *dynamically-changing* display mode, which allows for more complex transformations prior to or instead of the colormap. Possibly it is simply a matter of adding a new `image_transform` event and having ImageView subscribe to it.

Eventually, some of these transformations could be done on the GPU as part of a new VisPy Complex visual, but that seems like a longer-term project.

**Comment by pwinston** [comment]
I'm not sure I have the full context. But long term it seems like `ImageSlice` or something like `ImageSlice` should contain just the bytes loaded from disk/network. And then if Napari wants to display them some special way, that's a separate downstream step prior to display.

Like if we do have a cache some day, and your cache is full of chunks, and suddenly you change some display mode. We certainly don't want all those chunks to now be invalid, and we have the populate the cache again with new versions of that same data but that just look a little bit different. That would be a huge waste to reload that data.

Maybe that's exactly what you are saying above. But just stating it myself to see if I understand it.

**Comment by tlambert03** [comment]
> Eventually, some of these transformations could be done on the GPU as part of a new VisPy Complex visual, but that seems like a longer-term project.

@jni, I've got it mostly implemented on the vispy side.  Not GPU yet (started playing with it ... but that's heavy surgery since gl doesn't natively support GL_FLOAT_VEC2 textures)

**Comment by jni** [comment]
@pwinston That sounds right and pretty much what I meant. 

@tlambert03 My idea was that you could send .real and .imag to the GPU as two separate textures, then do the colormapping on the GPU based on those.

**Comment by pwinston** [comment]
Thinking about it you have the Model and the Viewer. But then I think we are saying the Viewer itself also has two layers. Kind of the raw layer and the cooked layer. Or maybe even more.

I think some layers already work like this. But I wonder if it needs to be formalized or standardized across the board. That every layer has this flow of model -> raw -> processed -> displayed. For `Image` the  `ImageView` class does part of that but it's very internal and ad hoc. Maybe there needs to be some type of processing pipeline or flow that layers can inject arbitrary processing into. Long long term.

I'm thinking of Photoshop adjustment layers and stuff like that. They do all sorts of weird processing steps that get kind of thrown into the stack of layers. Obviously all downstream of the actual image data. In that case the mix of dumb data and processing gets really complicated.

**Comment by d-v-b** [comment]
>Thinking about it you have the Model and the Viewer. But then I think we are saying the Viewer itself also has two layers. Kind of the raw layer and the cooked layer. Or maybe even more.

This goes back to an old "weird hill I wanted to die on", but... this issue could be addressed by taking a datatype-oriented approach. Imagine if napari dispatched off datatypes for array elements (like `uint8`, `uint16` (regular grayscale), but also `[uint8, uint8, uint8, uint8]` (RGBA), and maybe `[float64, float64]` (complex))... and each datatype determines the rendering modality. A datatype-driven approach can mitigate tons of "is this [10, 10, 4]-shaped array rgba or just small 3D" checks, because you can simply view the array as a [10, 10] array of rgba-typed elements, and then it's obvious how to render it. 

Without dispatching on array datatypes, i worry that the image layer is going to accumulate all this `is_$SAMPLE_TYPE` cruft that requires lots of special treatment before being sent off to vispy. But maybe there's some other solution I haven't thought of.

**Comment by pwinston** [comment]
So I think you are saying yes, the Viewer needs a raw data layer and a processed day layer. But then you are adding that the processing for different types `Image` layers is highly depending on the type of data. Such that we should have a clean way to map from datatype to processing code. And not have code that is constantly sniffing at the datatype and doing if/then branching.

Maybe what that means is two `Image` layers are not necessarily very similar at all. Instead what really matters is the data type that they contain.  In general maybe it means `Image` layers are generic containers of data, and until you know the type of the data you really can't do anything with them.

I wonder if this concept of containers vs. contents extends to non-image layers. Ideally with Napari we can generalize things that arise in one layer and apply them to other layers. We don't want to over do that, but we don't want every layer to re-invent the wheel, if in fact they are all tackling the same root issues.

The `Image` layer is special in a way, but we don't want it to be so special that other layers work totally differently.
